### ISSUE TITLE
Redis client tries to connect to slaves marked as down by Redis Sentinel

### ISSUE THREAD

**Issue by gregkare** [redis/redis-rb#827]
While replacing our Redis Sentinel setup we found an issue with the redis Ruby gem trying (and failing) to connect to a slave marked as down by Redis Sentinel (all of them had the `+sdown slave` message in their logs). I can reproduce the issue in a VM, with two slaves and a master (using Redis 4.10, 4.13 and 5.03 and the Ruby gem 3.2.1 and 4.1.0). So far I have found that raising the `reconnect_attempts` option of the Redis client results in successful connections to the slave that is up (we previously set it to 3, and 10 appears to be the sweet spot from what I have observed).

Is that a bug in the client library, or am I doing something wrong? If that's working as intended (Sentinels returning a slave that is known as down) I will open an issue in the Redis server repository

**Comment by supercaracal** [comment]
Hi,

I'd say that it might be the client side bug. The client internally calls `SENTINEL SLAVES master_name` command and fetches slave information but doesn't check flags.

https://github.com/redis/redis-rb/blob/v4.1.0/lib/redis/client.rb#L599-L607

I think we should check slave flags at the above step. `"flags"=>"s_down,slave,disconnected"`

I tried to reproduce the issue in local as follows.

| Redis | Port | remarks |
| --- | --- | --- |
| Sentinel 1 | 6400 |  |
| Sentinel 2 | 6401 |  |
| Sentinel 3 | 6402 |  |
| Master | 6381 |  |
| Slave 1 | 6382 |  |
| Slave 2 | 6383 | down |

```
$ bin/console

irb(main):001:0> r = Redis.new(host: 'master1', sentinels: [{ host: 'localhost', port: 6400 }], role: 'slave')
=> #<Redis client v4.1.0 for redis://master1:6379/0>

irb(main):002:0> r.ping
Redis::CannotConnectError (Error connecting to Redis on 127.0.0.1:6383 (Errno::ECONNREFUSED))

irb(main):003:0> Redis.new(port: 6400).sentinel(:slaves, 'master1')
=> [{"name"=>"127.0.0.1:6383", "ip"=>"127.0.0.1", "port"=>"6383", "runid"=>"436d285d1df3f7ef2a52522770464177f890298a", "flags"=>"s_down,slave,disconnected", "link-pending-commands"=>"3", "link-refcount"=>"1", "last-ping-sent"=>"1423549", "last-ok-ping-reply"=>"1423859", "last-ping-reply"=>"1423859", "s-down-time"=>"1418536", "down-after-milliseconds"=>"5000", "info-refresh"=>"1424093", "role-reported"=>"slave", "role-reported-time"=>"1946098", "master-link-down-time"=>"0", "master-link-status"=>"ok", "master-host"=>"127.0.0.1", "master-port"=>"6381", "slave-priority"=>"100", "slave-repl-offset"=>"101228"}, 
    {"name"=>"127.0.0.1:6382", "ip"=>"127.0.0.1", "port"=>"6382", "runid"=>"518ac95a1628ab495e93ca85a6fea7980488b096", "flags"=>"slave", "link-pending-commands"=>"0", "link-refcount"=>"1", "last-ping-sent"=>"0", "last-ok-ping-reply"=>"132", "last-ping-reply"=>"132", "down-after-milliseconds"=>"5000", "info-refresh"=>"7706", "role-reported"=>"slave", "role-reported-time"=>"1946106", "master-link-down-time"=>"0", "master-link-status"=>"ok", "master-host"=>"127.0.0.1", "master-port"=>"6381", "slave-priority"=>"100", "slave-repl-offset"=>"376206"}]

irb(main):004:0> r = Redis.new(host: 'master1', sentinels: [{ host: 'localhost', port: 6400 }], role: 'slave')
=> #<Redis client v4.1.0 for redis://master1:6379/0>

irb(main):005:0> r.ping
=> "PONG"

irb(main):006:0> r.connection
=> {:host=>"127.0.0.1", :port=>6382, :db=>0, :id=>"redis://127.0.0.1:6382/0", :location=>"127.0.0.1:6382"}
```

**Comment by byroot** [comment]
Fixed in #829
